### PR TITLE
Include JavaScript stack when Java throws.

### PR DESCRIPTION
**Pull request by szurbrigg**
- If a call to a bound Java method throws an exception, include intermediate JavaScript stack trace elements in the exception.

### CODE REVIEW

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java [hunk 1]
```diff
@@ -34,21 +34,36 @@
 
   public DuktapeException(String detailMessage) {
     super(getErrorMessage(detailMessage));
+    addDuktapeStack(this, detailMessage);
+  }
+
+  public static void addDuktapeStack(Throwable throwable, String detailMessage) {
```
Comment (swankjesse) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: package-private ?

File: tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java [hunk 2]
```diff
@@ -146,6 +146,57 @@ public String getValue() {
...
+          + "  value.getValue();\n"       // Line 10.
+          + "}\n", "test.js");
+      fail();
+    } catch (UnsupportedOperationException e) {
+      assertThat(e).hasMessage("Cannot getValue");
+
+      StackTraceElement[] stackTrace = e.getStackTrace();
+
+      // Top entry is what threw - the TestInterface implementation.
+      assertThat(stackTrace[0].getClassName()).isEqualTo(boundObject.getClass().getName());
```
Comment (swankjesse) [inline tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java]: might be more readable to just dump the stacktrace to a string & compare that? At a minimum that format is very familiar. 

```
  static String toStacktrace(Throwable t) {
    StringWriter writer = new StringWriter();
    t.printStackTrace(new PrintWriter(writer));
    return writer.toString();
  }
```

Comment (swankjesse) [inline tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java]: (though I guess the drawback of that is you also get the JUnit runtime stuff. So um, nevermind!)

File: tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java [hunk 3]
```diff
@@ -146,6 +146,57 @@ public String getValue() {
...
+      assertThat(stackTrace[4].getClassName()).isEqualTo(Duktape.class.getName());
+      assertThat(stackTrace[4].getMethodName()).isEqualTo("evaluate");
+      assertThat(stackTrace[4].isNativeMethod()).isTrue();
+
+      // Then the Java method.
+      assertThat(stackTrace[5].getClassName()).isEqualTo(Duktape.class.getName());
+      assertThat(stackTrace[5].getMethodName()).isEqualTo("evaluate");
+      assertThat(stackTrace[5].isNativeMethod()).isFalse();
+
+      // Then this test method.
```
Comment (swankjesse) [inline tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java]: Inception!

**Comment by swankjesse** [comment]
LGTM

File: duktape/src/main/java/com/squareup/duktape/DuktapeException.java (hunk 1 above)
Comment (szurbrigg) [inline duktape/src/main/java/com/squareup/duktape/DuktapeException.java]: Good call.  Done.

File: tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java (hunk 2 above)
Comment (szurbrigg) [inline tests/src/androidTest/java/com/squareup/duktape/DuktapeTest.java]: I had initially done that, and ran into just that.  There's also the ugliness/brittleness of the line numbers that show up in the printed stack trace.

**Pull request merged**